pkgdb/docs/database-locking.md

```diff
@@ -0,0 +1,133 @@
+- Keep waiting until you notice the file has disappeared.
+- Once the lock disappears you know the writer has finished.
+- You can now read the database.
+
+```mermaid
+sequenceDiagram
+    participant pkgdb
+    participant lock as abc123.lock
+    participant db as abc123.sqlite
+    pkgdb->>lock: Does it exist?
+    lock-->>pkgdb: Yes
+    pkgdb->>lock: Append PID to file
+    loop Watch the lock
+        pkgdb->>lock: Check mtime
+        lock-->>pkgdb: (the time)
+    end
+    pkgdb->>lock: Check mtime
+    lock-->>pkgdb: File not found
+    pkgdb->>db: Read
+```
+ 
+If the mtime of the lock is stale, the original writer must have crashed.
+At this point all waiting `pkgdb` processes read the PIDs in the lock.
+The first PID in the file takes over the writing process and removes its PID
+from the db lock so that if it crashes the next `pkgdb` process can take over.
+
+There are several possible race conditions in this procedure,
+namely "time of check vs. time of use,"
+so care is taken to mitigate them.
+The check whether the lock exists is atomic in that it is a call to `open` that
+fails if the file exists.
+
+Writing the PID to the lock first happens in a temporary file that is then
+renamed to the lock.
+This can overwrite a PID write from another process so each process writing its
+PID does a check afterwards to ensure that its PID is in the lock.
+If the PID isn't found in the lock,
+the read, update, write, rename process is repeated.
+This means that the final waiting order may not be the same order that the
+original interests were registered,
+but we actually don't care about that,
+we only care that one process writes while others wait and there's _some_ order
+for taking over writing if the writer fails.
+
+```mermaid
+sequenceDiagram
+    participant copyA as Temp A
+    participant pkgdbA as pkgdb A
+    participant lock as abc123.lock
+    participant pkgdbB as pkgdb B
+    participant copyB as Temp B
+    lock->>pkgdbA: Read PIDs
+    Note left of pkgdbA: Append pkgdb A's PID
+    pkgdbA->>copyA: Write PIDs
+    lock->>pkgdbB: Read PIDs
+    Note right of pkgdbB: Append pkgdb B's PID
+    pkgdbB->>copyB: Write PIDs
+    copyA->>lock: Rename
+    Note right of lock: pkgdb A's update is lost
+    copyB->>lock: Rename
+```
+
+## Consequences
+
+Since we shouldn't touch the database unless we have the lock,
+this means we can't automatically initialize the database
+(someone else might be writing to it) without the lock.
+
+An initial draft of this required you to acquire the lock and manually call
+`PkgDbInput::init`,
+but this turned out to be error prone.
+As a compromise between nice APIs, practicallity, and safety,
+the lock is now acquired whenever a `PkgDbInput` is created,
+which means that the lock is acquired twice for every database during scraping.
+This turns out to add negligible overhead,
+so it's not really an issue in practice,
+but those of us that care about the utmost efficiency may turn their noses up.
```

Comment:
nonblocking:
Thanks for adding docs, that was helpful.
- Why does there need to be some ordering for taking the lock? Could you just do an atomic open, and whoever wins wins?
- Are we locking when we just need readonly access? Is that just because we don't care too much about optimizing that? I think that's worth noting if it's deliberate. Also probably over-optimization at the moment, but if the toplevel attrSet of a DB is `done` we could stop locking that DB entirely.

Reply:
> Could you do an atomic open?

We kind of do this already. See the `O_EXCL` flag when opening the lock. You need this level of locking because if you let more than one person open the database then you'll get "database is locked" while doing a write.

> Are we locking only on read only access?

No, we're locking on every database access. The reason is that you could try to open a database for reading while someone else is still writing to it (constructing it). While launching several search process you could see this because one process would be constructing it and the next would see that a table hadn't been constructed yet. This locking also adds a very small amount of overhead (I measured it).

Reply:
> We kind of do this already. See the `O_EXCL` flag when opening the lock. You need this level of locking because if you let more than one person open the database then you'll get "database is locked" while doing a write.

I meant could you get rid of the whole PID queue. Just try to grab the lock whenever you wake up.

Reply:
> Are we locking only on read only access?

I'm not sure how you quote replied something I didn't write lol, but I think I understand. Once the section of the database we need is fully generated, we could avoid acquiring the lock, but for now acquire the lock for every process because it's low overhead.

Reply:
> I meant could you get rid of the whole PID queue. Just try to grab the lock whenever you wake up.

Maybe? I didn't try that. In general the race conditions are hard to deal with because if you accidentally let two processes acquire the lock one will delete the lock out from under the other one (that's what the troubleshooting section in the docs is about).

Reply:
I think it would be simpler to just have one race if there isn't a reason we need an ordering. Then you wouldn't have to do the whole rename from a temporary and double check your PID is still present dance. Nonblocking though

Reply:
I was thinking the following, but there's a race around removal:
```
if mtime lock has expired {
    remove file
    if try create lock with O_EXCL {
        I won
    } else {
        somebody else beat me
        sleep
    }
} else {
    sleep
}
```

I think a similar race is actually still present in what you have (unless I'm missing some synchronization around the double check):
A reads empty lock
B reads empty lock
A adds PID, renames to lock, double checks it's in the lock
B adds PID, renames to lock, double checks it's in the lock

Given @aakropotkin's comments though it might be better to just rely on SQLite locking mechanisms